src/linker/Linker.Dataflow/MethodBodyScanner.cs

```diff
@@ -894,5 +894,61 @@ private static VariableDefinition GetLocalDef (Instruction operation, Collection
 			Instruction operation,
 			ValueNodeList methodParams,
 			out ValueNode methodReturnValue);
+
+		private void ScanStelem (
+			Instruction operation,
+			Stack<StackSlot> currentStack,
+			MethodBody methodBody,
+			int curBasicBlock)
+		{
+			StackSlot valueToStore = PopUnknown (currentStack, 1, methodBody, operation.Offset);
+			StackSlot indexToStoreAt = PopUnknown (currentStack, 1, methodBody, operation.Offset);
+			StackSlot arrayToStoreIn = PopUnknown (currentStack, 1, methodBody, operation.Offset);
+			int? indexToStoreAtInt = indexToStoreAt.Value.AsConstInt ();
+			foreach (var array in arrayToStoreIn.Value.UniqueValues ()) {
+				if (array is ArrayValue arrValue) {
+					if (indexToStoreAtInt == null) {
+						// Since we can't know the current index we're storing the value at, clear all indices.
+						// That way we won't accidentally think we know the value at a given index when we cannot.
+						foreach (var knownIndex in arrValue.IndexValues.Keys) {
+							StoreMethodLocalValue (arrValue.IndexValues, UnknownValue.Instance, knownIndex, curBasicBlock);
+						}
+					} else {
+						// When we know the index, we can record the value at that index.
+						StoreMethodLocalValue (arrValue.IndexValues, valueToStore.Value, indexToStoreAtInt.Value, curBasicBlock);
+					}
+				}
+			}
+		}
+
+
+		private void ScanLdelem (
+			Instruction operation,
+			Stack<StackSlot> currentStack,
+			MethodBody methodBody)
+		{
+			StackSlot indexToLoadFrom = PopUnknown (currentStack, 1, methodBody, operation.Offset);
+			StackSlot arrayToLoadFrom = PopUnknown (currentStack, 1, methodBody, operation.Offset);
+			if (arrayToLoadFrom.Value is not ArrayValue arr) {
+				PushUnknown (currentStack);
+				return;
+			}
+			int? index = indexToLoadFrom.Value.AsConstInt ();
+			if (index == null) {
+				PushUnknown (currentStack);
```

Comment:
Should we also erase the whole array if this was ldelema? (This feels equivalent to a store with unknown index.)